src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +148,297 @@ public interface IFileWatchingToken
+                UnwatchDirectories,
+                UnwatchFiles,
+            }
+
+            /// <summary>
+            /// Represents a watcher operation that takes no action when applied. This value intentionally has the same
+            /// representation as <c>default(WatcherOperation)</c>.
+            /// </summary>
+            public static WatcherOperation Empty => new(Kind.None);
+
+            public static WatcherOperation WatchDirectory(string directory, string? filter, IVsFreeThreadedFileChangeEvents2 sink, List<uint> cookies)
+                => new(Kind.WatchDirectory, directory, filter, sink, cookies);
+
+            public static WatcherOperation WatchFile(string path, _VSFILECHANGEFLAGS fileChangeFlags, IVsFreeThreadedFileChangeEvents2 sink, Context.FileWatchingToken token)
+                => new(Kind.WatchFile, path, fileChangeFlags, sink, token);
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<Context.FileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(Context.FileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            /// <summary>
+            /// Attempts to combine the current <see cref="WatcherOperation"/> with the next operation in sequence. When
+            /// successful, <paramref name="combined"/> is assigned a value which, when applied, performs an operation
+            /// equivalent to performing the current instance immediately followed by <paramref name="other"/>.
+            /// </summary>
+            /// <param name="other">The next operation to apply.</param>
+            /// <param name="combined">An operation representing the combined application of the current instance and
+            /// <paramref name="other"/>, in that order; otherwise, <see cref="Empty"/> if the current operation cannot
+            /// be combined with <paramref name="other"/>.</param>
+            /// <returns><see langword="true"/> if the current operation can be combined with <paramref name="other"/>;
+            /// otherwise, <see langword="false"/>.</returns>
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                switch (_kind)
+                {
+                    case Kind.WatchDirectory:
+                    case Kind.WatchFile:
+                        // Watching operations cannot be combined
+                        break;
```

Comment:
interesting to explicitly have this.  don't love it, but dont' hate it.

Reply:
Keeping it mostly because it gave a clean place to document it.